Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -0,0 +1,969 @@
+
+                // Emit: var ae = task.Exception;
+                
+                var aeLocal = new Local(aggregateExceptionType.Value);
+
+                checkExceptionBlock.Statements.Add(
+                    new AssignmentStatement(aeLocal,
+                        new MethodCall(
+                            new MemberBinding(checkMethodTaskParameter,
+                                GetTaskProperty(checkMethodTaskParameter, "get_Exception")),
+                            new ExpressionList())));
+
+                // Emit: CheckException(ae);
+
+                // Need to store method result somewhere, otherwise stack would be corrupted
+                var checkResultLocal = new Local(SystemTypes.Boolean);
+
+                checkExceptionBlock.Statements.Add(
+                    new AssignmentStatement(checkResultLocal,
+                            new MethodCall(new MemberBinding(null, 
+                                checkExceptionMethod),
+                            new ExpressionList(checkExceptionMethod.ThisParameter, aeLocal))));
+
+                checkExceptionBlock.Statements.Add(endOfExceptionPrecondition);
+                
+                methodBody.Add(checkExceptionBlock);
+            }
+
+            // Previously, following line was executed only without exceptional postconditions!
+            // With new implementation this should be done for both: normal and exceptional postconditions.
+            // no exceptional post conditions
+            newBodyBlock.Statements.Add(origBody);
+
+            Block returnBlock = CreateReturnBlock(checkMethodTaskParameter, lastEnsuresSourceContext);
+            methodBody.Add(returnBlock);
+        }
+
+        private static IEnumerable<Ensures> GetTaskResultBasedEnsures(List<Ensures> asyncPostconditions)
+        {
+            foreach (var post in asyncPostconditions)
+            {
+                if (!(post is EnsuresExceptional))
+                {
+                    yield return post;
+                }
+            }
+        }
+
+        private static IEnumerable<EnsuresExceptional> GetExceptionalEnsures(List<Ensures> asyncPostconditions)
+        {
+            foreach (var post in asyncPostconditions)
+            {
+                var exceptional = post as EnsuresExceptional;
+                if (exceptional != null)
+                {
+                    yield return exceptional;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns TaskExtensions.Unwrap method.
+        /// </summary>
+        [Pure]
+        private static Member GetUnwrapMethod(TypeNode checkMethodTaskType)
+        {
+            Contract.Requires(checkMethodTaskType != null);
+            Contract.Ensures(Contract.Result<Member>() != null);
+
+            // TODO: use static for this?
+            var taskExtensions =
+                    HelperMethods.FindType(SystemCoreAssembly,
```

Comment:
:question: Will this still work for corefx, where `TaskExtensions` is not in System.Core? I don't think we need to address it now, but if we know it will fail for some cases that users will want it probably makes sense to add an issue to track it.